raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1593,28 +1587,61 @@ public void poll() throws IOException {
     }
 
     private void handlePendingAppends(long currentTimeMs) {
-        UnwrittenAppend unwrittenAppend = unwrittenAppends.poll();
-        if (unwrittenAppend == null || unwrittenAppend.isCancelled())
-            return;
-
         if (quorum.isLeader()) {
-            if (unwrittenAppend.isTimedOut(currentTimeMs)) {
-                unwrittenAppend.fail(new TimeoutException());
-            } else {
-                LeaderState leaderState = quorum.leaderStateOrThrow();
-                int epoch = quorum.epoch();
-                LogAppendInfo info = maybeAppendAsLeader(leaderState, unwrittenAppend.records);
-                if (info != null) {
-                    OffsetAndEpoch offsetAndEpoch = new OffsetAndEpoch(info.lastOffset, epoch);
-                    unwrittenAppend.complete(offsetAndEpoch);
-                    uncommittedAppends.put(offsetAndEpoch, new AppendBatchAndTime(info.lastOffset - info.firstOffset + 1, currentTimeMs));
+            int count = 0;
+            while (!unwrittenAppends.isEmpty() && count < MAX_APPENDS_PER_ITERATION) {
```

Comment:
An alternative would be to use the size of `unwrittenAppends` at the start of this loop as the maximum. Any appends that happen during the iteration will get postponed until the next iteration.